File: electron/ipc/update.ts
/**
 * 版本更新 IPC 处理器
 * 包装 electron-updater,提供手动模式(check / download / install)与事件推送
 */
import { app, ipcMain, BrowserWindow } from 'electron'
import { autoUpdater } from 'electron-updater'
import fs from 'node:fs'
import path from 'node:path'
import { IPC_CHANNELS } from '../../shared/ipc-channels'
import type {
  UpdateState,
  UpdateInfo,
  UpdateProgress,
  UpdateErrorPayload,
} from '../../shared/types'

interface Deps {
  getMainWindow: () => BrowserWindow | null
}

let currentState: UpdateState = 'idle'
let currentInfo: UpdateInfo | null = null

function setState(next: UpdateState): void {
  currentState = next
}

interface UpdateCache {
  lastCheckedAt: string
  lastVersion: string
}

function cacheFile(): string {
  return path.join(app.getPath('userData'), 'update-cache.json')
}

function loadCache(): UpdateCache | null {
  const f = cacheFile()
  if (!fs.existsSync(f)) return null
  try {
    return JSON.parse(fs.readFileSync(f, 'utf-8')) as UpdateCache
  } catch {
    return null
  }
}

function saveCache(cache: UpdateCache): void {
  fs.writeFileSync(cacheFile(), JSON.stringify(cache), 'utf-8')
}

function send<T>(channel: string, payload: T): void {
  const win = mainWindowGetter?.()
  win?.webContents.send(channel, payload)
}

let mainWindowGetter: Deps['getMainWindow'] | null = null

export function getCurrentState(): { state: UpdateState; info: UpdateInfo | null } {
  return { state: currentState, info: currentInfo }
}

export function registerHandlers(deps: Deps): void {
  mainWindowGetter = deps.getMainWindow

  // 事件订阅
  autoUpdater.on('update-available', (info) => {
    currentInfo = {
      version: info.version,
      releaseDate: info.releaseDate ?? '',
      releaseNotes: typeof info.releaseNotes === 'string' ? info.releaseNotes : '',
      sizeBytes: 0, // electron-updater 在 available 阶段不提供大小
    }
    setState('available')
    saveCache({ lastCheckedAt: new Date().toISOString(), lastVersion: info.version })
    send(IPC_CHANNELS.UPDATE_EVENT_AVAILABLE, currentInfo)
  })

  autoUpdater.on('update-not-available', () => {
    setState('idle')
    saveCache({ lastCheckedAt: new Date().toISOString(), lastVersion: '' })
    send(IPC_CHANNELS.UPDATE_EVENT_NOT_AVAILABLE, null)
  })

  autoUpdater.on('download-progress', (progress) => {
    const payload: UpdateProgress = {
      bytesPerSecond: progress.bytesPerSecond,
      percent: progress.percent,
      transferred: progress.transferred,
      total: progress.total,
    }
    send(IPC_CHANNELS.UPDATE_EVENT_PROGRESS, payload)
  })

  autoUpdater.on('update-downloaded', (info) => {
    if (currentInfo) {
      currentInfo = { ...currentInfo, version: info.version, sizeBytes: 0 }
    }
    setState('downloaded')
    send(IPC_CHANNELS.UPDATE_EVENT_DOWNLOADED, currentInfo)
  })

  autoUpdater.on('error', (err) => {
    const payload: UpdateErrorPayload = {
      code: 'unknown',
      message: err?.message ?? '未知错误',
    }
    send(IPC_CHANNELS.UPDATE_EVENT_ERROR, payload)
  })

  // 通道注册
  ipcMain.handle(IPC_CHANNELS.UPDATE_CHECK, async () => {
    if (!app.isPackaged) {
      // dev 模式跳过真实请求
      return { success: true, data: { skipped: true } }
    }
    try {
      setState('idle')
      await autoUpdater.checkForUpdates()
      return { success: true, data: { skipped: false } }
    } catch (err) {
      return { success: false, error: (err as Error).message }
    }
  })

  ipcMain.handle(IPC_CHANNELS.UPDATE_DOWNLOAD, async () => {
    try {
      setState('downloading')
      await autoUpdater.downloadUpdate()
      return { success: true, data: true }
    } catch (err) {
      setState('available') // 回滚到可重试状态
      return { success: false, error: (err as Error).message }
    }
  })

  ipcMain.handle(IPC_CHANNELS.UPDATE_INSTALL, () => {
    setState('installing')
    // quitAndInstall 是同步发起的;返回 success 仅表示发起成功
    setImmediate(() => autoUpdater.quitAndInstall())
    return { success: true, data: true }
  })

  ipcMain.handle(IPC_CHANNELS.UPDATE_GET_STATE, () => {
    return { success: true, data: getCurrentState() }
  })

  ipcMain.handle('update:get-last-check', () => {
    return { success: true, data: loadCache() }
  })
}

File: src-tauri/src/commands/update.rs
use std::sync::Mutex;

use tauri::{AppHandle, Emitter, Manager, State};
use tauri_plugin_updater::UpdaterExt;

use crate::models::dto::{
    IpcResult, UpdateCheckResult, UpdateErrorPayload, UpdateInfo, UpdateProgress, UpdateState,
};

// ─── Tauri event names (mapped from Electron IPC_CHANNELS.UPDATE_EVENT_*) ───
const EVENT_AVAILABLE: &str = "update://available";
const EVENT_NOT_AVAILABLE: &str = "update://not-available";
const EVENT_PROGRESS: &str = "update://progress";
const EVENT_DOWNLOADED: &str = "update://downloaded";
const EVENT_ERROR: &str = "update://error";

// ─── State management ───────────────────────────────────────────────────────

/// Managed state for the update lifecycle.
/// Mirrors the module-level `currentState` / `currentInfo` in electron/ipc/update.ts.
pub struct UpdateStateMgr(pub Mutex<UpdateStateInternal>);

pub struct UpdateStateInternal {
    pub state: UpdateState,
    pub update_info: Option<UpdateInfo>,
    pub error: Option<String>,
}

impl Default for UpdateStateMgr {
    fn default() -> Self {
        Self(Mutex::new(UpdateStateInternal {
            state: UpdateState::Idle,
            update_info: None,
            error: None,
        }))
    }
}

fn set_state(mgr: &UpdateStateMgr, new_state: UpdateState) {
    if let Ok(mut guard) = mgr.0.lock() {
        guard.state = new_state;
    }
}

fn set_info(mgr: &UpdateStateMgr, info: Option<UpdateInfo>) {
    if let Ok(mut guard) = mgr.0.lock() {
        guard.update_info = info;
    }
}

fn set_error(mgr: &UpdateStateMgr, error: Option<String>) {
    if let Ok(mut guard) = mgr.0.lock() {
        guard.error = error;
    }
}

fn emit_error(app: &AppHandle, message: String) {
    let _ = app.emit(
        EVENT_ERROR,
        UpdateErrorPayload {
            code: "unknown".into(),
            message,
        },
    );
}

// ─── Commands ───────────────────────────────────────────────────────────────

/// update:check — check for available updates via tauri-plugin-updater.
///
/// On success, pushes `update://available` (with UpdateInfo) or
/// `update://not-available` event to the frontend.
#[tauri::command]
pub async fn update_check(app: AppHandle) -> IpcResult<UpdateCheckResult> {
    let state = app.state::<UpdateStateMgr>();
    set_state(&state, UpdateState::Idle);
    set_error(&state, None);

    let updater = match app.updater() {
        Ok(u) => u,
        Err(e) => {
            let msg = e.to_string();
            set_error(&state, Some(msg.clone()));
            emit_error(&app, msg);
            return IpcResult::err("更新插件未配置");
        }
    };

    match updater.check().await {
        Ok(Some(update)) => {
            let info = UpdateInfo {
                version: update.version.clone(),
                release_date: update
                    .date
                    .map(|d| d.to_string())
                    .unwrap_or_default(),
                release_notes: update.body.clone().unwrap_or_default(),
                size_bytes: 0,
            };
            set_state(&state, UpdateState::Available);
            set_info(&state, Some(info.clone()));
            let _ = app.emit(EVENT_AVAILABLE, &info);
            IpcResult::ok(UpdateCheckResult {
                available: true,
                info: Some(info),
            })
        }
        Ok(None) => {
            let _ = app.emit(EVENT_NOT_AVAILABLE, ());
            IpcResult::ok(UpdateCheckResult {
                available: false,
                info: None,
            })
        }
        Err(e) => {
            let msg = e.to_string();
            set_error(&state, Some(msg.clone()));
            emit_error(&app, msg);
            IpcResult::err("检查更新失败")
        }
    }
}

/// update:download — download and install the update.
///
/// Uses `download_and_install` which streams the package and installs it.
/// Pushes `update://progress` events during download and
/// `update://downloaded` (with UpdateInfo) on completion.
#[tauri::command]
pub async fn update_download(app: AppHandle) -> IpcResult<bool> {
    let state = app.state::<UpdateStateMgr>();
    set_state(&state, UpdateState::Downloading);
    set_error(&state, None);

    let updater = match app.updater() {
        Ok(u) => u,
        Err(e) => {
            let msg = e.to_string();
            set_error(&state, Some(msg.clone()));
            set_state(&state, UpdateState::Available);
            emit_error(&app, msg);
            return IpcResult::err("更新插件未配置");
        }
    };

    let update = match updater.check().await {
        Ok(Some(u)) => u,
        Ok(None) => {
            set_state(&state, UpdateState::Idle);
            return IpcResult::err("无可用的更新");
        }
        Err(e) => {
            let msg = e.to_string();
            set_error(&state, Some(msg.clone()));
            set_state(&state, UpdateState::Available);
            emit_error(&app, msg);
            return IpcResult::err("检查更新失败");
        }
    };

    let info = UpdateInfo {
        version: update.version.clone(),
        release_date: update
            .date
            .map(|d| d.to_string())
            .unwrap_or_default(),
        release_notes: update.body.clone().unwrap_or_default(),
        size_bytes: 0,
    };
    set_info(&state, Some(info.clone()));

    let app_for_chunks = app.clone();
    let app_for_finish = app.clone();
    let mut transferred: u64 = 0;

    match update
        .download_and_install(
            move |chunk_len, content_length| {
                transferred += chunk_len as u64;
                let percent = content_length
                    .map(|total| (transferred as f64 / total as f64) * 100.0)
                    .unwrap_or(0.0);
                let _ = app_for_chunks.emit(
                    EVENT_PROGRESS,
                    UpdateProgress {
                        bytes_per_second: 0.0,
                        percent,
                        transferred,
                        total: content_length.unwrap_or(0),
                    },
                );
            },
            move || {
                let _ = app_for_finish.emit(EVENT_DOWNLOADED, &info);
            },
        )
        .await
    {
        Ok(()) => {
            set_state(&state, UpdateState::Downloaded);
            IpcResult::ok(true)
        }
        Err(e) => {
            let msg = e.to_string();
            set_error(&state, Some(msg.clone()));
            set_state(&state, UpdateState::Available);
            emit_error(&app, msg);
            IpcResult::err("下载更新失败")
        }
    }
}

/// update:install — mark state as Installing.
///
/// The actual installation is performed by `download_and_install` in update_download.
/// App relaunch is handled by the frontend via @tauri-apps/plugin-process.
#[tauri::command]
pub async fn update_install(app: AppHandle) -> IpcResult<bool> {
    if let Some(mgr) = app.try_state::<UpdateStateMgr>() {
        set_state(&mgr, UpdateState::Installing);
    }
    IpcResult::ok(true)
}

/// update:get-state — return the current update state.
#[tauri::command]
pub fn update_get_state(state: State<UpdateStateMgr>) -> IpcResult<UpdateState> {
    match state.0.lock() {
        Ok(guard) => IpcResult::ok(guard.state.clone()),
        Err(e) => IpcResult::err(e.to_string()),
    }
}
